### PR TITLE
PE-46 Add backend pipelines and update frontend

### PR DESCRIPTION
**Pull request by ElfoLiNk**
- Add deploy and code-review pipelines for [cgn-onboarding-portal-backend](https://github.com/pagopa/cgn-onboarding-portal-backend)
- Update deploy pipeline for [cgn-onboarding-portal-frontend](https://github.com/pagopa/cgn-onboarding-portal-frontend)

### CODE REVIEW

File: azure-devops/projects/cgn-onboarding-portal-projects/secrets.tf [hunk 1]
```diff
@@ -12,6 +12,10 @@ variable "secrets" {
     "io-azure-devops-github-USERNAME",
     "pagopa-npm-bot-TOKEN",
     "TTDIO-SPN-TENANTID",
+    "TTDIO-UAT-GCNPORTAL-SUBSCRIPTION-ID",
+    "TTDIO-PROD-GCNPORTAL-SUBSCRIPTION-ID",
```
Comment (pasqualedevita) [inline azure-devops/projects/cgn-onboarding-portal-projects/secrets.tf:16]:
```suggestion
    "PAGOPAIT-TENANTID",
    "PAGOPAIT-UAT-GCNPORTAL-SUBSCRIPTION-ID",
    "PAGOPAIT-PROD-GCNPORTAL-SUBSCRIPTION-ID",
```
I think that npm connection is not needed
*-CGNPORTAL are on another tenant, pagopa.it

File: azure-devops/projects/cgn-onboarding-portal-projects/secrets.tf [hunk 2]
```diff
@@ -12,6 +12,10 @@ variable "secrets" {
...
+    "cgnportal-sonarqube-URL",
+    "cgnportal-sonarqube-TOKEN"
```
Comment (pasqualedevita) [inline azure-devops/projects/cgn-onboarding-portal-projects/secrets.tf:18]:
```suggestion
    "cgnportal-sonarqube-URL",
    "cgnportal-sonarqube-TOKEN"
```
Is it your sonarqube or sonarcloud?

File: azure-devops/projects/cgn-onboarding-portal-projects/service_connections.tf [hunk 3]
```diff
@@ -1,27 +1,27 @@
...
+# Azure service connection PROD-GCNPORTAL
+resource "azuredevops_serviceendpoint_azurerm" "PROD-GCNPORTAL" {
+  depends_on = [azuredevops_project.project]
+
+  project_id                = azuredevops_project.project.id
+  service_endpoint_name     = "PROD-GCNPORTAL-SERVICE-CONN"
+  description               = "PROD-GCNPORTAL Service connection"
+  azurerm_subscription_name = "PROD-GCNPORTAL"
+  azurerm_spn_tenantid      = data.azurerm_key_vault_secret.key_vault_secret["TTDIO-SPN-TENANTID"].value
+  azurerm_subscription_id   = data.azurerm_key_vault_secret.key_vault_secret["TTDIO-PROD-GCNPORTAL-SUBSCRIPTION-ID"].value
```
Comment (pasqualedevita) [inline azure-devops/projects/cgn-onboarding-portal-projects/service_connections.tf:10]:
```suggestion
  azurerm_spn_tenantid      = data.azurerm_key_vault_secret.key_vault_secret["PAGOPAIT-TENANTID"].value
  azurerm_subscription_id   = data.azurerm_key_vault_secret.key_vault_secret["PAGOPAIT-PROD-GCNPORTAL-SUBSCRIPTION-ID"].value
```

File: azure-devops/projects/cgn-onboarding-portal-projects/service_connections.tf [hunk 4]
```diff
@@ -1,27 +1,27 @@
...
+# Azure service connection UAT-GCNPORTAL
+resource "azuredevops_serviceendpoint_azurerm" "UAT-GCNPORTAL" {
+  depends_on = [azuredevops_project.project]
+
+  project_id                = azuredevops_project.project.id
+  service_endpoint_name     = "UAT-GCNPORTAL-SERVICE-CONN"
+  description               = "UAT-GCNPORTAL Service connection"
+  azurerm_subscription_name = "UAT-GCNPORTAL"
+  azurerm_spn_tenantid      = data.azurerm_key_vault_secret.key_vault_secret["TTDIO-SPN-TENANTID"].value
+  azurerm_subscription_id   = data.azurerm_key_vault_secret.key_vault_secret["TTDIO-UAT-GCNPORTAL-SUBSCRIPTION-ID"].value
```
Comment (pasqualedevita) [inline azure-devops/projects/cgn-onboarding-portal-projects/service_connections.tf:22]:
```suggestion
  azurerm_spn_tenantid      = data.azurerm_key_vault_secret.key_vault_secret["PAGOPAIT-TENANTID"].value
  azurerm_subscription_id   = data.azurerm_key_vault_secret.key_vault_secret["PAGOPAIT-UAT-GCNPORTAL-SUBSCRIPTION-ID"].value
```

File: azure-devops/projects/cgn-onboarding-portal-projects/service_connections.tf [hunk 5]
```diff
@@ -65,18 +65,43 @@ resource "azuredevops_serviceendpoint_github" "io-azure-devops-github-pr" {
   }
 }
 
-# Github service connection (read-only) used to access to azure-pipelines template repo
-resource "azuredevops_serviceendpoint_github" "pagopa" {
+# azure container registry service connection
+resource "azuredevops_serviceendpoint_azurecr" "cgnonboardingportal_uat_azurecr" {
```
Comment (pasqualedevita) [inline azure-devops/projects/cgn-onboarding-portal-projects/service_connections.tf:69]:
```suggestion
resource "azuredevops_serviceendpoint_azurecr" "cgnonboardingportal-uat-azurecr" {
```
it's more correct use _ but other resources use -

File: azure-devops/projects/cgn-onboarding-portal-projects/service_connections.tf [hunk 6]
```diff
@@ -65,18 +65,43 @@ resource "azuredevops_serviceendpoint_github" "io-azure-devops-github-pr" {
...
+  depends_on = [azuredevops_project.project]
+
+  project_id                = azuredevops_project.project.id
+  service_endpoint_name     = "cgnonboardingportal-uat-azurecr"
+  resource_group            = "cgnonboardingportal-u-api-rg"
+  azurecr_name              = "cgnonboardingportaluarc"
+  azurecr_subscription_name = "UAT-Esercenti"
+  azurecr_spn_tenantid      = data.azurerm_key_vault_secret.key_vault_secret["TTDIO-SPN-TENANTID"].value
+  azurecr_subscription_id   = data.azurerm_key_vault_secret.key_vault_secret["TTDIO-UAT-GCNPORTAL-SUBSCRIPTION-ID"].value
```
Comment (pasqualedevita) [inline azure-devops/projects/cgn-onboarding-portal-projects/service_connections.tf:78]:
```suggestion
  azurecr_spn_tenantid      = data.azurerm_key_vault_secret.key_vault_secret["PAGOPAIT-TENANTID"].value
  azurecr_subscription_id   = data.azurerm_key_vault_secret.key_vault_secret["PAGOPAIT-UAT-GCNPORTAL-SUBSCRIPTION-ID"].value
```

**Review by pasqualedevita — CHANGES_REQUESTED** [review]
thank you @ElfoLiNk I added some suggestions, also for front-end u need to add a "Storage Blob Data Contributor" role assignment to PROD/UAT service connection.

You can see an example here https://github.com/pagopa/gitops/blob/main/azure-devops/projects/io-pay-projects/io-pay.tf#L210

File: azure-devops/projects/cgn-onboarding-portal-projects/secrets.tf (hunk 2 above)
Comment (ElfoLiNk) [reply azure-devops/projects/cgn-onboarding-portal-projects/secrets.tf:17]: Should be this one https://github.com/pagopa/cgn-onboarding-portal-infra/blob/master/code/sonarqube.tf#L39

cgnportal-sonarqube-URL should be "https://cgnonboardingportal-u-sonarqube.westeurope.azurecontainer.io/"

It's not yet deployed in UAT due to enable_sonarqube being false

**Comment by ElfoLiNk** [comment]
> thank you @ElfoLiNk I added some suggestions, also for front-end u need to add a "Storage Blob Data Contributor" role assignment to PROD/UAT service connection.
> 
> You can see an example here https://github.com/pagopa/gitops/blob/main/azure-devops/projects/io-pay-projects/io-pay.tf#L210

Working on it, Thanks.

File: azure-devops/projects/cgn-onboarding-portal-projects/service_connections.tf [hunk 7]
```diff
@@ -1,27 +1,47 @@
...
-//
-//  project_id                = azuredevops_project.project.id
-//  service_endpoint_name     = "DEV-IO-SERVICE-CONN"
-//  description               = "DEV-IO Service connection"
-//  azurerm_subscription_name = "DEV-IO"
-//  azurerm_spn_tenantid      = data.azurerm_key_vault_secret.key_vault_secret["TTDIO-SPN-TENANTID"].value
-//  azurerm_subscription_id   = data.azurerm_key_vault_secret.key_vault_secret["TTDIO-DEV-IO-SUBSCRIPTION-ID"].value
-//}
+provider "azuread" {
+  tenant_id = data.azurerm_key_vault_secret.key_vault_secret["TTDIO-SPN-TENANTID"].value
```
Comment (pasqualedevita) [inline azure-devops/projects/cgn-onboarding-portal-projects/service_connections.tf:2]:
```suggestion
  tenant_id = data.azurerm_key_vault_secret.key_vault_secret["PAGOPA-TENANTID"].value
```

**Review by pasqualedevita — CHANGES_REQUESTED** [review]

File: azure-devops/projects/cgn-onboarding-portal-projects/service_connections.tf [hunk 8]
```diff
@@ -1,27 +1,47 @@
...
+provider "azuread" {
+  tenant_id = data.azurerm_key_vault_secret.key_vault_secret["PAGOPAIT-TENANTID"].value
+}
+
+locals {
+  PROD-GCNPORTAL-UID = "${local.azure_devops_org}-${azuredevops_project.project.name}-${data.azurerm_key_vault_secret.key_vault_secret["PAGOPAIT-PROD-GCNPORTAL-SUBSCRIPTION-ID"].value}"
+  UAT-GCNPORTAL-UID  = "${local.azure_devops_org}-${azuredevops_project.project.name}-${data.azurerm_key_vault_secret.key_vault_secret["PAGOPAIT-UAT-GCNPORTAL-SUBSCRIPTION-ID"].value}"
+  service_principal_uids = [
+    local.PROD-GCNPORTAL-UID,
+    local.PROD-GCNPORTAL-UID,
```
Comment (pasqualedevita) [inline azure-devops/projects/cgn-onboarding-portal-projects/service_connections.tf:10]:
```suggestion
    local.UAT-GCNPORTAL-UID,
```

**Review by pasqualedevita — CHANGES_REQUESTED** [review]
also add these required providers (azuread is missing)

```
  required_providers {
    azuredevops = {
      source  = "microsoft/azuredevops"
      version = "= 0.1.3"
    }
    azurerm = {
      version = "~> 2.52.0"
    }
    azuread = {
      source  = "hashicorp/azuread"
      version = "1.4.0"
    }
    time = {
      version = "~> 0.6.0"
    }
  }
```

File: azure-devops/projects/cgn-onboarding-portal-projects/cgn-onboarding-portal-frontend.tf [hunk 9]
```diff
@@ -164,12 +207,38 @@ resource "azuredevops_resource_authorization" "cgn-onboarding-portal-frontend-de
...
 //  type          = "endpoint"
 //}
+
+resource "azurerm_role_assignment" "cgn-onboarding-portal-frontend-deploy-azurerm-PROD-IO-storageaccount" {
+  depends_on = [data.azuread_service_principal.service_principals]
+
+  principal_id         = data.azuread_service_principal.service_principals[local.PROD-GCNPORTAL-UID].id
+  role_definition_name = "Storage Blob Data Contributor"
+  scope                = "/subscriptions/${data.azurerm_key_vault_secret.key_vault_secret["PAGOPAIT-PROD-GCNPORTAL-SUBSCRIPTION-ID"].value}/resourceGroups/${var.cgn-onboarding-portal-frontend.pipeline.prod.resource_group_azure}/providers/Microsoft.Storage/storageAccounts/${var.cgn-onboarding-portal-frontend.pipeline.prod.storage_account_name}"
+}
```
Comment (pasqualedevita) [inline azure-devops/projects/cgn-onboarding-portal-projects/cgn-onboarding-portal-frontend.tf:226]:
```suggestion
# resource "azurerm_role_assignment" "cgn-onboarding-portal-frontend-deploy-azurerm-PROD-GCNPORTAL-storageaccount" {
#   depends_on = [data.azuread_service_principal.service_principals]

#   principal_id         = data.azuread_service_principal.service_principals[local.PROD-GCNPORTAL-UID].id
#   role_definition_name = "Storage Blob Data Contributor"
#   scope                = "/subscriptions/${data.azurerm_key_vault_secret.key_vault_secret["PAGOPAIT-PROD-GCNPORTAL-SUBSCRIPTION-ID"].value}/resourceGroups/${var.cgn-onboarding-portal-frontend.pipeline.prod.resource_group_azure}/providers/Microsoft.Storage/storageAccounts/${var.cgn-onboarding-portal-frontend.pipeline.prod.storage_account_name}"
# }
```

**Review by pasqualedevita — CHANGES_REQUESTED** [review]

File: azure-devops/projects/cgn-onboarding-portal-projects/cgn-onboarding-portal-backend.tf [hunk 10]
```diff
@@ -0,0 +1,176 @@
+variable "cgn-onboarding-portal-backend" {
+  default = {
+    repository = {
+      organization   = "pagopa"
+      name           = "cgn-onboarding-portal-backend"
+      branch_name    = "master"
+      pipelines_path = ".devops"
+    }
+    pipeline = {
+      # TODO ?
```
Comment (pasqualedevita) [inline azure-devops/projects/cgn-onboarding-portal-projects/cgn-onboarding-portal-backend.tf:10]:
```suggestion
      # TODO ?
      prod = {
        webAppName = "cgnonboardingportal-p-portal-backend1"
      }
      uat = {
        webAppName = "cgnonboardingportal-u-portal-backend1"
      }
```

**Review by pasqualedevita — COMMENTED** [review]
@ElfoLiNk what do u think?

File: azure-devops/projects/cgn-onboarding-portal-projects/cgn-onboarding-portal-frontend.tf [hunk 11]
```diff
@@ -164,12 +207,38 @@ resource "azuredevops_resource_authorization" "cgn-onboarding-portal-frontend-de
...
+  depends_on = [azuredevops_serviceendpoint_azurerm.UAT-GCNPORTAL, azuredevops_build_definition.cgn-onboarding-portal-frontend-deploy, time_sleep.wait]
+
+  project_id    = azuredevops_project.project.id
+  resource_id   = azuredevops_serviceendpoint_azurerm.UAT-GCNPORTAL.id
+  definition_id = azuredevops_build_definition.cgn-onboarding-portal-frontend-deploy.id
+  authorized    = true
+  type          = "endpoint"
+}
+
+resource "azurerm_role_assignment" "cgn-onboarding-portal-frontend-deploy-azurerm-UAT-GCN-storageaccount" {
```
Comment (pasqualedevita) [inline azure-devops/projects/cgn-onboarding-portal-projects/cgn-onboarding-portal-frontend.tf:238]:
```suggestion
resource "azurerm_role_assignment" "cgn-onboarding-portal-frontend-deploy-azurerm-UAT-GCNPORTAL-storageaccount" {
```

**Review by pasqualedevita — CHANGES_REQUESTED** [review]

File: azure-devops/projects/cgn-onboarding-portal-projects/cgn-onboarding-portal-backend.tf (hunk 10 above)
Comment (ElfoLiNk) [reply azure-devops/projects/cgn-onboarding-portal-projects/cgn-onboarding-portal-backend.tf]: Yes, if we need to rename the web app we don't need to change the application pipeline

**Comment by pasqualedevita** [comment]
@AleDore can u add pagopa-bot as admin on cgn-portal-backend repo?

**Review by pasqualedevita — APPROVED** [review]
thanks @ElfoLiNk applied

**Comment by pasqualedevita** [comment]
test pr ok!!!

**Pull request merged**